File: src/main/java/jisop/command_line/ControllerRecognizer.java
package jisop.command_line;

import jisop.MissingArgumentException;
import jisop.command_line.lex.ArgumentLexer;
import jisop.command_line.lex.Token;
import jisop.command_line.parse.ArgumentParameter;
import jisop.command_line.parse.ArgumentParser;
import jisop.command_line.parse.ArgumentWithOptions;
import jisop.command_line.parse.ParsedArguments;
import jisop.domain.*;

import java.util.Arrays;
import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

/**
 *
 * @author mathieu
 */
public class ControllerRecognizer {

    private final boolean _allowInferParameter;
    private final Configuration _configuration;
    private final Controller _controller;
    private final Function<Class,Object> _typeContainer;

    public ControllerRecognizer(Controller controller, Configuration configuration, Function<Class, Object> typeContainer, boolean allowInferParameter) {
        this._controller = controller;
        this._configuration = configuration;
        this._typeContainer = typeContainer;
        this._allowInferParameter = allowInferParameter;
    }

    public ParsedArguments parseArgumentsAndMerge(String actionName, Map<String, String> arg, ParsedArguments parsedArguments) {
        Method methodInfo = _controller.getMethod(actionName);
        Collection<Argument> argumentRecognizers = methodInfo.getArguments();

        ArgumentParser parser = new ArgumentParser(argumentRecognizers, _allowInferParameter);
        ParsedArguments parsedMethodArguments = parser.parse(arg);
        ParsedMethod parsedMethod = parse(methodInfo, parsedMethodArguments);
        ParsedArguments merged = parsedArguments.merge(parsedMethod);
        if (!_controller.ignoreGlobalUnMatchedParameters)
            merged.assertFailOnUnMatched();
        return merged;
    }

    public ParsedArguments parseArgumentsAndMerge(String[] arg, ParsedArguments parsedArguments) {
        ParsedArguments parsedMethod = parse(arg);
        // Inferred ordinal arguments should not be recognized twice
        parsedArguments.recognizedArguments = parsedArguments.recognizedArguments
                .stream()
                .filter(argopts ->
                !parsedMethod.recognizedArguments.stream()
                        .anyMatch(pargopt -> pargopt.index == argopts.index && argopts.inferredOrdinal))
                .collect(Collectors.toList())
        ;
        ParsedArguments merged = parsedArguments.merge(parsedMethod);
        if (!_controller.ignoreGlobalUnMatchedParameters)
            merged.assertFailOnUnMatched();
        return merged;
    }

    public boolean recognize(String[] arg) {
        List<Token> lexed = RewriteLexedTokensToSupportHelpAndIndex.rewrite(ArgumentLexer.lex(arg));
        return null != findMethodInfo(lexed);
    }

    public List<Argument> getRecognizers(String method) {
        return _controller.getMethod(method).getArguments();
    }

    /// <summary>
    /// Note that in order to register a converter you can use:
    /// TypeDescriptor.AddAttributes(typeof(AType), new TypeConverterAttribute(typeof(ATypeConverter)));
    /// </summary>
    /// <param name="arg"></param>
    /// <returns></returns>
    public ParsedMethod parse(String[] arg) {
        List<Token> lexed = RewriteLexedTokensToSupportHelpAndIndex.rewrite(ArgumentLexer.lex(arg));

        Method methodInfo = findMethodInfo(lexed);

        List<Argument> argumentRecognizers = methodInfo.getArguments();
        argumentRecognizers.addAll(0, Arrays.asList(
                new ArgumentWithOptions(ArgumentParameter.parse("#0" + _controller.name)).type(String.class).required(true),
                new ArgumentWithOptions(ArgumentParameter.parse("#1" + methodInfo.name)).required(false).type(String.class))
        );

        ArgumentParser parser = new ArgumentParser(argumentRecognizers, _allowInferParameter);
        ParsedArguments parsedArguments = parser.parse(lexed, Arrays.asList(arg));

        return parse(methodInfo, parsedArguments);
    }

    private Method findMethodInfo(List<Token> arg) {
        if (arg.size()<2){
            return null;
        }
        boolean foundClassName = _controller.name.equalsIgnoreCase(arg.get(0).value);
        if (foundClassName)
        {
            String methodName = arg.get(1).value;
            return FindMethodAmongLexedTokens.findMethod(
                    _controller.getControllerActionMethods(),
                    methodName, arg);
        }
        return null;
    }

    public ParsedMethod parse(Method methodInfo, ParsedArguments parsedArguments)
    {
        Collection<Argument> unMatchedRequiredArguments = parsedArguments.UnMatchedRequiredArguments();
        if (unMatchedRequiredArguments.stream().anyMatch(a -> true))
        {
            throw new MissingArgumentException("Missing arguments",
                    unMatchedRequiredArguments.stream().map(a->a.name).toArray(String[]::new))
            ;
        }
        ConvertArgumentsToParameterValue convertArgument = new ConvertArgumentsToParameterValue(_configuration.typeConverter);
        Collection<Object> recognizedActionParameters = convertArgument.getParametersForMethod(methodInfo,
                parsedArguments.RecognizedArgumentsAsPairs(),
                _typeContainer);
        ParsedMethod m= new ParsedMethod( parsedArguments, _typeContainer, _configuration);
        m.recognizedAction = methodInfo;
        m.recognizedActionParameters = recognizedActionParameters.toArray();
        m.recognizedClass = _controller.type;
        return m;
    }

}

File: src/main/java/isop/command_line/ControllerRecognizer.kt
package isop.command_line

import isop.MissingArgumentException
import isop.command_line.lex.ArgumentLexer
import isop.command_line.lex.Token
import isop.command_line.parse.ArgumentParameter
import isop.command_line.parse.ArgumentParser
import isop.command_line.parse.ArgumentWithOptions
import isop.command_line.parse.ParsedArguments
import isop.domain.*

import java.util.Arrays
import java.util.function.Function
import java.util.stream.Collectors

/**
 *
 * @author mathieu
 */
class ControllerRecognizer(private val _controller: Controller, private val _configuration: Configuration, private val _typeContainer: Function<Class<*>, Any?>, private val _allowInferParameter: Boolean) {

    fun parseArgumentsAndMerge(actionName: String, arg: Map<String, String>, parsedArguments: ParsedArguments): ParsedArguments {
        val methodInfo = _controller.getMethod(actionName)
        val argumentRecognizers = methodInfo!!.arguments

        val parser = ArgumentParser(argumentRecognizers, _allowInferParameter)
        val parsedMethodArguments = parser.parse(arg)
        val parsedMethod = parse(methodInfo, parsedMethodArguments)
        val merged = parsedArguments.merge(parsedMethod)
        if ((!_controller.ignoreGlobalUnMatchedParameters!!)!!)
            merged.assertFailOnUnMatched()
        return merged
    }

    fun parseArgumentsAndMerge(arg: Array<String>, parsedArguments: ParsedArguments): ParsedArguments {
        val parsedMethod = parse(arg)
        // Inferred ordinal arguments should not be recognized twice
        parsedArguments.recognizedArguments = parsedArguments.recognizedArguments
                .filter { argopts ->
                    !parsedMethod.recognizedArguments.stream()
                            .anyMatch { pargopt -> pargopt.index == argopts.index && argopts.inferredOrdinal }
                }
                .toList()
        val merged = parsedArguments.merge(parsedMethod)
        if ((!_controller.ignoreGlobalUnMatchedParameters!!)!!)
            merged.assertFailOnUnMatched()
        return merged
    }

    fun recognize(arg: Array<String>): Boolean {
        val lexed = RewriteLexedTokensToSupportHelpAndIndex.rewrite(ArgumentLexer.lex(arg))
        return null != findMethodInfo(lexed)
    }

    fun getRecognizers(method: String): List<Argument> {
        return _controller.getMethod(method)!!.arguments
    }

    /// <summary>
    /// Note that in order to register a converter you can use:
    /// TypeDescriptor.AddAttributes(typeof(AType), new TypeConverterAttribute(typeof(ATypeConverter)));
    /// </summary>
    /// <param name="arg"></param>
    /// <returns></returns>
    fun parse(arg: Array<String>): ParsedMethod {
        val lexed = RewriteLexedTokensToSupportHelpAndIndex.rewrite(ArgumentLexer.lex(arg))

        val methodInfo = findMethodInfo(lexed)

        val argumentRecognizers = methodInfo!!.arguments.toMutableList()
        argumentRecognizers.addAll(0, Arrays.asList<ArgumentWithOptions>(
                ArgumentWithOptions(ArgumentParameter.parse("#0" + _controller.name)).type(String::class.java).required(true),
                ArgumentWithOptions(ArgumentParameter.parse("#1" + methodInfo.name)).required(false).type(String::class.java))
        )

        val parser = ArgumentParser(argumentRecognizers, _allowInferParameter)
        val parsedArguments = parser.parse(lexed, Arrays.asList(*arg))

        return parse(methodInfo, parsedArguments)
    }

    private fun findMethodInfo(arg: List<Token>): Method? {
        if (arg.size < 2) {
            return null
        }
        val foundClassName = _controller.name.equals(arg[0].value, ignoreCase = true)
        if (foundClassName) {
            val methodName = arg[1].value
            return FindMethodAmongLexedTokens.findMethod(
                    _controller.controllerActionMethods,
                    methodName, arg)
        }
        return null
    }

    fun parse(methodInfo: Method, parsedArguments: ParsedArguments): ParsedMethod {
        val unMatchedRequiredArguments = parsedArguments.UnMatchedRequiredArguments()
        if (unMatchedRequiredArguments.stream().anyMatch { a -> true }) {
            throw MissingArgumentException("Missing arguments",
                    unMatchedRequiredArguments.map { a -> a.name }.toTypedArray())
        }
        val convertArgument = ConvertArgumentsToParameterValue(_configuration.typeConverter)
        val recognizedActionParameters = convertArgument.getParametersForMethod(methodInfo,
                parsedArguments.RecognizedArgumentsAsPairs(),
                _typeContainer)
        val m = ParsedMethod(parsedArguments,
                _controller.type,
                methodInfo,recognizedActionParameters.toTypedArray(),
                _typeContainer, _configuration)
        return m
    }

}
